File: automaton-maui/Core/Validator.cs
using AutomatonDesigner.Models;

namespace AutomatonDesigner.Core;

/// <summary>Structural validation for automata.</summary>
public static class Validator
{
    public static List<ValidationIssue> Validate(Automaton aut)
    {
        var issues = new List<ValidationIssue>();
        issues.AddRange(CheckStartState(aut));
        issues.AddRange(CheckAcceptStates(aut));
        issues.AddRange(CheckTransitionStates(aut));
        issues.AddRange(CheckTransitionSymbols(aut));
        issues.AddRange(CheckUnreachableStates(aut));
        issues.AddRange(CheckDeterminism(aut));
        return issues;
    }

    public static bool IsValid(Automaton aut) => Validate(aut).Count == 0;

    static IEnumerable<ValidationIssue> CheckStartState(Automaton aut)
    {
        if (!aut.States.Contains(aut.Start))
            yield return new StartStateNotInStates(aut.Start);
    }

    static IEnumerable<ValidationIssue> CheckAcceptStates(Automaton aut) =>
        aut.Accept.Where(s => !aut.States.Contains(s)).Select(s => new AcceptStateNotInStates(s));

    static IEnumerable<ValidationIssue> CheckTransitionStates(Automaton aut) =>
        aut.Transitions.SelectMany(t =>
        {
            var issues = new List<ValidationIssue>();
            if (!aut.States.Contains(t.From)) issues.Add(new TransitionFromUndefinedState(t));
            if (!aut.States.Contains(t.To)) issues.Add(new TransitionToUndefinedState(t));
            return issues;
        });

    static IEnumerable<ValidationIssue> CheckTransitionSymbols(Automaton aut) =>
        aut.Transitions
            .Where(t => t.Label is OnSymbol s && !aut.Alphabet.Contains(s.Symbol))
            .Select(t => new UndefinedSymbolInTransition(((OnSymbol)t.Label).Symbol, t));

    static IEnumerable<ValidationIssue> CheckUnreachableStates(Automaton aut)
    {
        var reachable = ComputeReachable(aut);
        return aut.States
            .Where(s => s != aut.Start && !reachable.Contains(s))
            .Select(s => new UnreachableState(s));
    }

    static IEnumerable<ValidationIssue> CheckDeterminism(Automaton aut)
    {
        if (aut.Type != AutomatonType.DFA) yield break;
        foreach (var state in aut.States)
        {
            foreach (var sym in aut.Alphabet)
            {
                int count = aut.Transitions.Count(t =>
                    t.From == state && t.Label is OnSymbol s && s.Symbol == sym);
                if (count > 1)
                    yield return new NonDeterministicTransition(state, sym);
            }
        }
    }

    static HashSet<State> ComputeReachable(Automaton aut)
    {
        var visited = new HashSet<State> { aut.Start };
        var frontier = new HashSet<State> { aut.Start };
        while (frontier.Count > 0)
        {
            var next = new HashSet<State>();
            foreach (var t in aut.Transitions)
            {
                if (frontier.Contains(t.From) && !visited.Contains(t.To))
                    next.Add(t.To);
            }
            if (next.Count == 0) break;
            visited.UnionWith(next);
            frontier = next;
        }
        return visited;
    }
}

File: automaton-designer/src/Automaton/Validator.hs
{-# LANGUAGE OverloadedStrings #-}

-- | Structural validation for automata.
--
-- Checks for unreachable states, undefined symbols in transitions,
-- states referenced in transitions but not declared, determinism
-- violations (for DFA), and other structural issues.
module Automaton.Validator
  ( validate
  , isValid
  , formatIssue
  ) where

import Data.Set  (Set)
import qualified Data.Set  as Set
import Data.Text (Text)
import qualified Data.Text as T

import Automaton.Types

-- ---------------------------------------------------------------------------
-- Public API
-- ---------------------------------------------------------------------------

-- | Validate an automaton and return all structural issues found.
validate :: Automaton -> [ValidationIssue]
validate aut = concat
  [ checkStartState aut
  , checkAcceptStates aut
  , checkTransitionStates aut
  , checkTransitionSymbols aut
  , checkUnreachableStates aut
  , checkDeterminism aut
  ]

-- | Check whether an automaton is free of structural issues.
isValid :: Automaton -> Bool
isValid = null . validate

-- | Format a 'ValidationIssue' as a human-readable message.
formatIssue :: ValidationIssue -> Text
formatIssue (UnreachableState s) =
  "Unreachable state: " <> stateName s
formatIssue (UndefinedSymbolInTransition sym tr) =
  "Symbol '" <> T.singleton (symbolChar sym) <> "' in transition "
  <> fmtTrans tr <> " is not in the alphabet"
formatIssue (DuplicateTransition tr) =
  "Duplicate transition: " <> fmtTrans tr
formatIssue MissingStartState =
  "No start state defined"
formatIssue (StartStateNotInStates s) =
  "Start state " <> stateName s <> " is not in the set of states"
formatIssue (AcceptStateNotInStates s) =
  "Accept state " <> stateName s <> " is not in the set of states"
formatIssue (TransitionFromUndefinedState tr) =
  "Transition from undefined state: " <> fmtTrans tr
formatIssue (TransitionToUndefinedState tr) =
  "Transition to undefined state: " <> fmtTrans tr
formatIssue (NonDeterministicTransition s sym) =
  "Non-deterministic: state " <> stateName s
  <> " has multiple transitions on '" <> T.singleton (symbolChar sym) <> "'"

fmtTrans :: Transition -> Text
fmtTrans (Transition from lbl to') =
  stateName from <> "," <> fmtLabel lbl <> " -> " <> stateName to'
  where
    fmtLabel (OnSymbol s) = T.singleton (symbolChar s)
    fmtLabel OnEpsilon    = "ε"

-- ---------------------------------------------------------------------------
-- Individual checks
-- ---------------------------------------------------------------------------

-- | Start state must be declared in the set of states.
checkStartState :: Automaton -> [ValidationIssue]
checkStartState aut
  | automStart aut `Set.member` automStates aut = []
  | otherwise = [StartStateNotInStates (automStart aut)]

-- | All accept states must be declared.
checkAcceptStates :: Automaton -> [ValidationIssue]
checkAcceptStates aut =
  [ AcceptStateNotInStates s
  | s <- Set.toList (automAccept aut)
  , not (s `Set.member` automStates aut)
  ]

-- | Source and target states of every transition must be declared.
checkTransitionStates :: Automaton -> [ValidationIssue]
checkTransitionStates aut = concatMap check (Set.toList (automTransitions aut))
  where
    states = automStates aut
    check t = concat
      [ [ TransitionFromUndefinedState t | not (transFrom t `Set.member` states) ]
      , [ TransitionToUndefinedState t   | not (transTo   t `Set.member` states) ]
      ]

-- | Every symbol used in a transition must be in the alphabet.
checkTransitionSymbols :: Automaton -> [ValidationIssue]
checkTransitionSymbols aut =
  [ UndefinedSymbolInTransition s t
  | t <- Set.toList (automTransitions aut)
  , OnSymbol s <- [transLabel t]
  , not (s `Set.member` automAlphabet aut)
  ]

-- | Find states that are not reachable from the start state.
checkUnreachableStates :: Automaton -> [ValidationIssue]
checkUnreachableStates aut =
  [ UnreachableState s
  | s <- Set.toList (automStates aut)
  , s /= automStart aut
  , not (s `Set.member` reachable)
  ]
  where
    reachable = computeReachable aut

-- | For DFAs, each (state, symbol) pair must have at most one transition.
checkDeterminism :: Automaton -> [ValidationIssue]
checkDeterminism aut
  | automType aut /= DFA = []
  | otherwise =
    [ NonDeterministicTransition s sym
    | s   <- Set.toList (automStates aut)
    , sym <- Set.toList (automAlphabet aut)
    , let targets =
            [ transTo t
            | t <- Set.toList (automTransitions aut)
            , transFrom t == s
            , transLabel t == OnSymbol sym
            ]
    , length targets > 1
    ]

-- ---------------------------------------------------------------------------
-- Reachability helper
-- ---------------------------------------------------------------------------

-- | Compute the set of states reachable from the start state
-- via any transition (BFS).
computeReachable :: Automaton -> Set State
computeReachable aut = go initVisited initVisited
  where
    initVisited = Set.singleton (automStart aut)
    go visited frontier
      | Set.null newStates = visited
      | otherwise          = go visited' newStates
      where
        reachable = Set.fromList
          [ transTo t
          | t <- Set.toList (automTransitions aut)
          , transFrom t `Set.member` frontier
          ]
        newStates = Set.difference reachable visited
        visited'  = Set.union visited newStates
